lib/form/paragraph3.flow

```diff
@@ -630,160 +655,139 @@ keepFormTogether(form : Form) -> bool {
 		);
 	});
 
-	Pair(
-		Baseline(const(a.baseline), Size2(const(WidthHeight(if (tightWidth) a.width else max(availableWidth, a.width), a.y)),
-			Group(list2array(a.lines))
-		)),
-		a.width
+	MeasuredForm(
+		Baseline(
+			const(a.baseline),
+			Size2(const(WidthHeight(width, a.y)), Group(list2array(a.lines))),
+		),
+		WidthHeight(longestLineWidth, a.y)
 	);
 }
 
-ParaLineResult(form : Form, width : double, height : double, baseline : double);
+ParaLineMetrics(width : double, height : double, asc : double);
 OptimizedLineElement(f : GhostForm, metrics : Behaviour<FormMetrics>, interactivityIdM : Maybe<int>);
-
-getOptimizedLine(words : [ParaWord], alignment : ParaLineAlignment) -> [OptimizedLineElement]{
-	if (!isBiDiEnabled() && alignment != Justify()) {
-		optimizeLine(words);
-	} else {
-		map(words, \w -> OptimizedLineElement(w.form, w.metrics, w.interactivityIdM));
-	};
+OptimizedLine(words : [OptimizedLineElement], metrics : ParaLineMetrics, indent : double);
+
+getOptimizedLine(line : ParaLine, justify : bool) -> OptimizedLine {
+	words = if (isBiDiEnabled() || justify)
+		map(line.words, \w -> OptimizedLineElement(w.form, w.metrics, w.interactivityIdM))
+	else
+		optimizeLine(line.words);
+	lineAsc = fold(words, 0.0, \ac, w -> max(ac, getValue(w.metrics).baseline));
+	wh = fold(words, WidthHeight(line.indent, 0.0), \ac, w -> WidthHeight(
+		ac.width + getValue(w.metrics).width,
+		max(ac.height, lineAsc + getValue(w.metrics).height - getValue(w.metrics).baseline)
+	));
+	OptimizedLine(words, ParaLineMetrics(wh.width, wh.height, lineAsc), line.indent)
 }
 
 // Aligns all elements to the baseline, as well as calculates the height and the baseline
 // Send out the result, width, height
 RenderLine(
 	y : double,
-	words : [ParaWord],
+	optimizedLine : OptimizedLine,
 	width : double,
-	tightWidth : bool,
-	alignment : ParaLineAlignment,
-	lineIndent: double,
+	alignment : Maybe<double>,
 	lineNumber : int,
 	interactiveStyles : Tree<int, [ParaElementInteractiveStyle]>,
 	rtl : bool,
 	isMarked
-) -> ParaLineResult {
-	optimizedLine = getOptimizedLine(words, alignment);
-	metrics = map(optimizedLine, \p -> getValue(p.metrics));
-	lineWidth = lineIndent + fold(metrics, 0.0, \ac, m -> ac + m.width); // this is real width of the line at the moment.
-	lineAsc = fold(metrics, 0.0, \ac, m -> max(ac, m.baseline));
-	lineHeight = fold(metrics, 0.0, \ac, m -> max(ac, lineAsc + m.height - m.baseline));
-
+) -> Form {
 	// Alignment offset
-	startOffset = if (rtl) max(width - lineWidth, 0.) else 0.;
-	endOffset = if (rtl) 0. else max(0., width - lineWidth);
-	alignmentOffset = ref switch (alignment : ParaLineAlignment) {
-		LeftAlign(): if (rtl) endOffset else startOffset;
-		RightAlign(): if (rtl) startOffset else endOffset;
-		CenterAlign(): max(0.0, (width - lineWidth) / 2.0);
-		StartAlign(): startOffset;
-		EndAlign(): endOffset;
-		Justify(): 0.0;
-	} + (if (rtl) 0. else lineIndent);
+	dWidth = max(0.0, width - optimizedLine.metrics.width);
```

Comment:
please, use good naming convention.
dWidth and so on is not good naming

Reply:
d is good for difference or delta. Why make the name twice longer while d represents the idea?

Reply:
Finally renamed as it was before.